staging/src/k8s.io/apiserver/pkg/admission/interfaces.go

```diff
@@ -60,6 +60,12 @@ type Attributes interface {
 	// An error is returned if the format of key is invalid. When trying to overwrite annotation with a new value, an error is returned.
 	// Both ValidationInterface and MutationInterface are allowed to add Annotations.
 	AddAnnotation(key, value string) error
+
+	// TODO(mehdy): Doc
+	GetObjectConverter() runtime.ObjectConvertor
+
+	// TODO(mehdy): Doc
+	GetObjectCreator() runtime.ObjectCreater
```

Comment:
admission uses the defaulter as well... this seems like it will converge to the set of `runtime.Object*` interfaces held by request scope over time

Reply:
I think a struct of all of these interfaces make more sense as we would need to pass a lot of parameters (and many nils in tests) if we don't do so. Something like a `type ObjectInterfaces struct {'. We can add interfaces to that as needed. Or we can just give access to the RequestScope here. I will read code to see which level of encapsulation make sense here. Recommendations are welcome.

Reply:
> I think a struct of all of these interfaces make more sense as we would need to pass a lot of parameters (and many nils in tests) if we don't do so. Something like a `type ObjectInterfaces struct {'.

I agree gathering them makes sense.

What if we made the wrapper an interface:
```go
type ObjectInterfaces interface {
	GetObjectCreater() runtime.ObjectCreater
	GetObjectTyper() runtime.ObjectTyper
	GetObjectDefaulter() runtime.ObjectDefaulter
	GetObjectConvertor() runtime.ObjectConvertor
}
```

which RequestScope could implement:
```go
func (r *RequestScope) GetObjectCreater() runtime.ObjectCreater     { return r.Creater }
func (r *RequestScope) GetObjectTyper() runtime.ObjectTyper         { return r.Typer }
func (r *RequestScope) GetObjectDefaulter() runtime.ObjectDefaulter { return r.Defaulter }
func (r *RequestScope) GetObjectConvertor() runtime.ObjectConvertor { return r.Convertor }
```

and could be passed alongside attributes to admit/validate methods:
```go
type MutationInterface interface {
	Interface

	// Admit makes an admission decision based on the request attributes
	Admit(a Attributes, i ObjectInterfaces) (err error)
}

// ValidationInterface is an abstract, pluggable interface for Admission Control decisions.
type ValidationInterface interface {
	Interface

	// Validate makes an admission decision based on the request attributes.  It is NOT allowed to mutate
	Validate(a Attributes, i ObjectInterfaces) (err error)
}
```

that keeps a clearer separation between the attributes being validated and the tools admission can make use of to do the validation

Reply:
I'm not thrilled about the signature changes, but I don't think we have any great options:
* put new items into a separate parameter to Admit/Validate
  * pro: separates data and helpers
  * pro: keeps interface of admission.Attributes the same
  * con: changes interface of Admit/Validate (impacts admission plugin implementers)
* put new items into admission.Attributes
  * pro: keeps Admit/Validate interfaces the same
  * con: mixes data being validated and helpers used to validate
  * con: changes interface of admission.Attributes (impacts admission plugin implementers)
* let plugins implement an alternate interface (`Admit2`) for getting called with helpers
  * pro: no interface changes
  * con: wrapping/unioning/chaining/delegation layers make it likely the alternate method would get masked
* let attributes implement an alternate interface interested admission plugins can attempt to cast it to to extract object helpers
  * pro: no interface changes
  * con: this is insane